packages/server-core/src/user/identity-provider/identity-provider.hooks.ts

```diff
@@ -44,56 +52,138 @@ import {
-    return context
   }
 }
 
-const checkOnlyIdentityProvider = () => {
-  return async (context: HookContext): Promise<HookContext> => {
-    if (!context.id) {
-      // do not allow to remove identity providers in bulk
-      throw new MethodNotAllowed('Cannot remove multiple providers together')
-    }
-    const thisIdentityProvider = (await context.app
-      .service(identityProviderPath)
-      .get(context.id)) as IdentityProviderType
+/** (BEFORE) CREATE HOOKS **/
 
-    if (!thisIdentityProvider) throw new Forbidden('You do not have any identity provider')
+// async function validateAuthParams(context:HookContext<IdentityProviderService>) {
+async function validateAuthParams(context: HookContext) {
+  let userId = context.data.userId
 
-    // we only want to disallow removing the last identity provider if it is not a guest
-    // since the guest user will be destroyed once they log in
-    if (thisIdentityProvider.type === 'guest') return context
+  if (context.params.authentication) {
+    const authResult = await (context.app.service('authentication') as any).strategies.jwt.authenticate(
+      { accessToken: context.params.authentication.accessToken },
+      {}
+    )
+    userId = userId || authResult[appConfig.authentication.entity]?.userId
+  }
 
-    const providers = await context.app
-      .service(identityProviderPath)
-      .find({ query: { userId: thisIdentityProvider.userId } })
+  if (!userId) {
+    return
+  }
+
+  const user = await context.app.service(userPath).get(userId)
+
+  context.existingUser = user
+}
 
-    if (providers.total <= 1) {
-      throw new MethodNotAllowed('Cannot remove the only identity provider on a user')
+async function addIdentityProviderType(context: HookContext) {
+  const isAdmin = context.existingUser && (await checkScope(context.existingUser, 'admin', 'admin'))
+  if (!isAdmin && context.params!.provider && !['password', 'email', 'sms'].includes(context.data.type)) {
+    context.data.type = 'guest'
+  }
+
+  const adminScopes = await context.app.service(scopePath).find({
+    query: {
+      type: 'admin:admin'
     }
-    return context
+  })
+
+  if (adminScopes.total === 0 && isDev && context.data.type !== 'guest') {
+    // doubt here - it should be the first if condition
+    context.data.type = 'admin'
+  }
+}
+
+async function createNewUser(context: HookContext) {
+  const isGuest = context.data.type === 'guest'
+  const avatars = await context.app.service(avatarPath).find({ isInternal: true, query: { $limit: 1000 } } as any)
+
+  const newUser = await context.app.service(userPath).create({
+    isGuest,
+    avatarId: avatars.data[random(avatars.data.length - 1)].id
+  })
+
+  context.existingUser = newUser
+}
+
+/** (AFTER) CREATE HOOKS **/
+
+async function addScopes(context: HookContext) {
+  if (context.data.type === 'guest' && appConfig.scopes.guest.length) {
```

Comment:
We should use `set-loggedin-user-in-body` and `set-loggedin-user-in-query` hooks instead of another custom hook

Reply:
implemented!